src/fu-engine.c

```diff
@@ -1759,22 +1759,29 @@ fu_engine_install_blob (FuEngine *self,
 	/* mark this as modified even if we actually fail to do the update */
 	fu_device_set_modified (device, (guint64) g_get_real_time () / G_USEC_PER_SEC);
 
-	/* signal to all the plugins the update is about to happen */
+	/* plugins can set FWUPD_DEVICE_FLAG_NEEDS_RETRY to run again, but they
+	 * must return TRUE rather than an error */
 	device_id = g_strdup (fu_device_get_id (device));
-	if (!fu_engine_update_prepare (self, flags, device_id, error))
-		return FALSE;
+	do {
+		/* don't rely on a plugin clearing this */
+		fu_device_remove_flag (device, FWUPD_DEVICE_FLAG_NEEDS_RETRY);
 
-	/* detach to bootloader mode */
-	if (!fu_engine_update_detach (self, device_id, error))
-		return FALSE;
+		/* signal to all the plugins the update is about to happen */
+		if (!fu_engine_update_prepare (self, flags, device_id, error))
+			return FALSE;
 
-	/* install */
-	if (!fu_engine_update (self, device_id, blob_fw, flags, error))
-		return FALSE;
+		/* detach to bootloader mode */
+		if (!fu_engine_update_detach (self, device_id, error))
+			return FALSE;
 
-	/* attach into runtime mode */
-	if (!fu_engine_update_attach (self, device_id, error))
-		return FALSE;
+		/* install */
+		if (!fu_engine_update (self, device_id, blob_fw, flags, error))
+			return FALSE;
+
+		/* attach into runtime mode */
+		if (!fu_engine_update_attach (self, device_id, error))
+			return FALSE;
+	} while (fu_device_has_flag (device, FWUPD_DEVICE_FLAG_NEEDS_RETRY));
```

Comment:
I think a debugging line with a counter may be useful at this point

Reply:
Counter as in how many different loops we've done? Maybe we abort with an error after say 5?

Reply:
Yeah that's what I was thinking, to avoid daemon deadlock. Perhaps device will need to declare how many writes it needs though too then.